katib-ui-rock/rockcraft.yaml

```
@@ -19,36 +19,86 @@ platforms:
   amd64:
 
 parts:
-  katib-ui:
+  frontend-lib:
+    plugin: nil
+    source: https://github.com/kubeflow/kubeflow
+    source-type: git
+    # Set source-tag here to the commit cited in https://github.com/kubeflow/katib/blob/master/pkg/ui/v1beta1/frontend/COMMIT
+    # for this Katib release
+    # Note: Katib 0.16.0 cannot build with Kubeflow v1.8-branch because Kubeflow refactored this lib
+    source-commit: 046c6d3c8301a95a2809c0d927bbbf72638dd639
+    # Omit source-depth because we're cloning a commit
+    # source-depth: 1
+    build-environment:
+      # suppress analytics collection
+      - NG_CLI_ANALYTICS: false
+    build-snaps:
+      - node/12/stable
+    override-build: |
+      # Clean the build dir so we don't accidentally use files we didn't intend
+      rm -rf ./*
+
+      # Pull only the files we need for the build from src, and build
+      cp $CRAFT_PART_SRC/components/crud-web-apps/common/frontend/kubeflow-common-lib/package*.json ./
+      npm ci
+      cp -r $CRAFT_PART_SRC/components/crud-web-apps/common/frontend/kubeflow-common-lib/* ./
+      npm run build
+      
+    override-stage: |
+      mkdir -p $CRAFT_STAGE/frontend-lib/dist
+      cp -r $CRAFT_PART_BUILD/dist/kubeflow $CRAFT_STAGE/frontend-lib/dist/kubeflow
+
+  frontend:
+    after: [frontend-lib]
+    plugin: nil
+    source: https://github.com/kubeflow/katib
+    source-type: git
+    source-tag: v0.16.0
+    source-depth: 1
+    build-snaps:
+      - node/12/stable
+    override-build: |
+      # Clean the build dir so we don't accidentally use files we didn't intend
+      rm -rf ./*
+
+      # Pull only the files we need for the build from src
+      cp $CRAFT_PART_SRC/pkg/ui/v1beta1/frontend/package*.json ./
+      npm ci
+      cp -r $CRAFT_STAGE/frontend-lib/dist/kubeflow ./node_modules/kubeflow
+      cp -r $CRAFT_PART_SRC/pkg/ui/v1beta1/frontend/* .
```

nit but would it make sense to copy those the other way around in order make sure `node_modules/kubeflow` is intact? (you can always just resolve :smile: )

---

Sorry I don't understand what you mean.  Could you give an example?

---

```suggestion
      cp -r $CRAFT_PART_SRC/pkg/ui/v1beta1/frontend/* .
      cp -r $CRAFT_STAGE/frontend-lib/dist/kubeflow ./node_modules/kubeflow
```
That's my suggestion.

In order to explain my comment more, `$CRAFT_PART_SRC/pkg/ui/v1beta1/frontend/*` could include in theory a `.../frontend/node_modules/kubeflow/` folder. We copy everything and `node_modules` folder is being created during `npm ci` which has already run. Thus it would overwrite the previously copied `node_modules/kubeflow`. It struck me mainly because [upstream](https://github.com/kubeflow/katib/blob/v0.16.0/cmd/ui/v1beta1/Dockerfile#L32-L33) makes those copies with a different order. In practice that won't happen since katib-ui doesn't install any `kubeflow` module/package. But still, following what upstream does would protect us from this edge case scenario.

---

ohhh yeah totally agree good catch! I've updated that now